// investec-api/dist/util/investec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getInvestecCardMerchants = exports.getInvestecCardCurrencies = exports.getInvestecCardCountries = exports.postInvestecCardEnvironmentVariables = exports.getInvestecCardEnvironmentVariables = exports.getInvestecCardExecutions = exports.postInvestecSimulateExecuteFunctionCode = exports.postInvestecCardPublishSavedCode = exports.postInvestecCardSaveCode = exports.getInvestecCardPublishedCode = exports.getInvestecCardSavedCode = exports.getInvestecCards = exports.postInvestecTransferMultiple = exports.getInvestecTransactionsForAccount = exports.getAccountBalance = exports.getInvestecAccounts = exports.getInvestecToken = void 0;
const node_fetch_1 = require("node-fetch");
const getBasicHeaders = (token) => {
    return {
        Authorization: `Bearer ${token}`,
    };
};
const safeResponse = (response) => {
    if (response.status !== 200) {
        return { status: response.status };
    }
    return response.json();
};
const getInvestecToken = (clientId, clientSecret) => __awaiter(void 0, void 0, void 0, function* () {
    const tokenResponse = yield (0, node_fetch_1.default)("https://openapi.investec.com/identity/v2/oauth2/token", {
        method: "POST",
        body: `grant_type=client_credentials&client_id=${clientId}&client_secret=${clientSecret}&scope=accounts`,
        headers: {
            "Content-Type": "application/x-www-form-urlencoded",
        },
    });
    return safeResponse(tokenResponse);
});
exports.getInvestecToken = getInvestecToken;
const getInvestecAccounts = (token) => __awaiter(void 0, void 0, void 0, function* () {
    const accountsResponse = yield (0, node_fetch_1.default)(`https://openapi.investec.com/za/pb/v1/accounts`, {
        headers: Object.assign({}, getBasicHeaders(token)),
    });
    return safeResponse(accountsResponse);
});
exports.getInvestecAccounts = getInvestecAccounts;
const getAccountBalance = (token, accountId) => __awaiter(void 0, void 0, void 0, function* () {
    const balanceResponse = yield (0, node_fetch_1.default)(`https://openapi.investec.com/za/pb/v1/accounts/${accountId}/balance`, {
        headers: Object.assign({}, getBasicHeaders(token)),
    });
    return safeResponse(balanceResponse);
});
exports.getAccountBalance = getAccountBalance;
const getInvestecTransactionsForAccount = (token, { accountId, fromDate, toDate, transactionType, }) => __awaiter(void 0, void 0, void 0, function* () {
    const transactionsResponse = yield (0, node_fetch_1.default)(`https://openapi.investec.com/za/pb/v1/accounts/${accountId}/transactions?${fromDate ? ` &fromDate=${fromDate}` : ""}${toDate ? `&toDate=${toDate}` : ""}
      ${transactionType ? `&transactionType=${transactionType}` : ""}`, {
        headers: Object.assign({}, getBasicHeaders(token)),
    });
    return safeResponse(transactionsResponse);
});
exports.getInvestecTransactionsForAccount = getInvestecTransactionsForAccount;
const postInvestecTransferMultiple = (token, { fromAccountId, toAccounts, }) => __awaiter(void 0, void 0, void 0, function* () {
    const body = {
        AccountId: fromAccountId,
        TransferList: toAccounts.map((t) => ({
            BeneficiaryAccountId: t.accountId,
            Amount: t.amount,
            MyReference: t.myReference,
            TheirReference: t.theirReference,
        })),
    };
    const transferResponse = yield (0, node_fetch_1.default)("https://openapi.investec.com/za/pb/v1/accounts/transfermultiple", {
        method: "POST",
        body: JSON.stringify(body),
        headers: Object.assign({ "Content-Type": "application/json" }, getBasicHeaders(token)),
    });
    return safeResponse(transferResponse);
});
exports.postInvestecTransferMultiple = postInvestecTransferMultiple;
const getInvestecCards = (token) => __awaiter(void 0, void 0, void 0, function* () {
    const cardsResponse = yield (0, node_fetch_1.default)(`https://openapi.investec.com/za/v1/cards`, {
        headers: Object.assign({}, getBasicHeaders(token)),
    });
    return safeResponse(cardsResponse);
});
exports.getInvestecCards = getInvestecCards;
const getInvestecCardSavedCode = (token, cardKey) => __awaiter(void 0, void 0, void 0, function* () {
    const cardsResponse = yield (0, node_fetch_1.default)(`https://openapi.investec.com/za/v1/cards/${cardKey}/code`, {
        headers: Object.assign({}, getBasicHeaders(token)),
    });
    return safeResponse(cardsResponse);
});
exports.getInvestecCardSavedCode = getInvestecCardSavedCode;
const getInvestecCardPublishedCode = (token, cardKey) => __awaiter(void 0, void 0, void 0, function* () {
    const cardsResponse = yield (0, node_fetch_1.default)(`https://openapi.investec.com/za/v1/cards/${cardKey}/publishedcode`, {
        headers: Object.assign({}, getBasicHeaders(token)),
    });
    return safeResponse(cardsResponse);
});
exports.getInvestecCardPublishedCode = getInvestecCardPublishedCode;
const postInvestecCardSaveCode = (token, cardKey, code) => __awaiter(void 0, void 0, void 0, function* () {
    const body = { code };
    const response = yield (0, node_fetch_1.default)(`https://openapi.investec.com/za/v1/cards/${cardKey}/code`, {
        method: "POST",
        body: JSON.stringify(body),
        headers: Object.assign({ "Content-Type": "application/json" }, getBasicHeaders(token)),
    });
    return safeResponse(response);
});
exports.postInvestecCardSaveCode = postInvestecCardSaveCode;
const postInvestecCardPublishSavedCode = (token, cardKey, codeId) => __awaiter(void 0, void 0, void 0, function* () {
    const body = { codeid: codeId, code: "" };
    const response = yield (0, node_fetch_1.default)(`https://openapi.investec.com/za/v1/cards/${cardKey}/code`, {
        method: "POST",
        body: JSON.stringify(body),
        headers: Object.assign({ "Content-Type": "application/json" }, getBasicHeaders(token)),
    });
    return safeResponse(response);
});
exports.postInvestecCardPublishSavedCode = postInvestecCardPublishSavedCode;
const postInvestecSimulateExecuteFunctionCode = (token, cardKey, opts) => __awaiter(void 0, void 0, void 0, function* () {
    const body = Object.assign({}, opts);
    const response = yield (0, node_fetch_1.default)(`https://openapi.investec.com/za/v1/cards/${cardKey}/code/execute`, {
        method: "POST",
        body: JSON.stringify(body),
        headers: Object.assign({ "Content-Type": "application/json" }, getBasicHeaders(token)),
    });
    return safeResponse(response);
});
exports.postInvestecSimulateExecuteFunctionCode = postInvestecSimulateExecuteFunctionCode;
const getInvestecCardExecutions = (token, cardKey) => __awaiter(void 0, void 0, void 0, function* () {
    const cardsResponse = yield (0, node_fetch_1.default)(`https://openapi.investec.com/za/v1/cards/${cardKey}/code/executions`, {
        headers: Object.assign({}, getBasicHeaders(token)),
    });
    return safeResponse(cardsResponse);
});
exports.getInvestecCardExecutions = getInvestecCardExecutions;
const getInvestecCardEnvironmentVariables = (token, cardKey) => __awaiter(void 0, void 0, void 0, function* () {
    const envResponse = yield (0, node_fetch_1.default)(`https://openapi.investec.com/za/v1/cards/${cardKey}/environmentvariables`, {
        headers: Object.assign({}, getBasicHeaders(token)),
    });
    return safeResponse(envResponse);
});
exports.getInvestecCardEnvironmentVariables = getInvestecCardEnvironmentVariables;
const postInvestecCardEnvironmentVariables = (token, cardKey, variables) => __awaiter(void 0, void 0, void 0, function* () {
    const body = { variables };
    const response = yield (0, node_fetch_1.default)(`https://openapi.investec.com/za/v1/cards/${cardKey}/environmentvariables`, {
        method: "POST",
        body: JSON.stringify(body),
        headers: Object.assign({ "Content-Type": "application/json" }, getBasicHeaders(token)),
    });
    return safeResponse(response);
});
exports.postInvestecCardEnvironmentVariables = postInvestecCardEnvironmentVariables;
const getInvestecCardCountries = (token) => __awaiter(void 0, void 0, void 0, function* () {
    const envResponse = yield (0, node_fetch_1.default)(`https://openapi.investec.com/za/v1/cards/countries`, {
        headers: Object.assign({}, getBasicHeaders(token)),
    });
    return safeResponse(envResponse);
});
exports.getInvestecCardCountries = getInvestecCardCountries;
const getInvestecCardCurrencies = (token) => __awaiter(void 0, void 0, void 0, function* () {
    const envResponse = yield (0, node_fetch_1.default)(`https://openapi.investec.com/za/v1/cards/currencies`, {
        headers: Object.assign({}, getBasicHeaders(token)),
    });
    return safeResponse(envResponse);
});
exports.getInvestecCardCurrencies = getInvestecCardCurrencies;
const getInvestecCardMerchants = (token) => __awaiter(void 0, void 0, void 0, function* () {
    const envResponse = yield (0, node_fetch_1.default)(`https://openapi.investec.com/za/v1/cards/merchants`, {
        headers: Object.assign({}, getBasicHeaders(token)),
    });
    return safeResponse(envResponse);
});
exports.getInvestecCardMerchants = getInvestecCardMerchants;
